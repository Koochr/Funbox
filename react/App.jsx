/*global ymaps*/
import React from "react"
import SortableList from "./components/SortableList.jsx"
import InputForm from "./components/InputForm.jsx"

export default class App extends React.Component {
  state = {
  	wayPoints: [],
  	yaMap: {},
  }

  init = () => {
  	this.setState({
  		yaMap: new ymaps.Map("map", {
  			center: [55.76, 37.64,],
  			zoom: 10,
  		}),
  	})
  }

  refactorWayPoints() {
  	this.state.yaMap.geoObjects.removeAll()
  	for (let i=0; i<this.state.wayPoints.length; i++) {

    	let preset
  		switch (i) {
    		case 0:
    			preset = "islands#blueHomeCircleIcon"
  			break
  		case this.state.wayPoints.length-1:
  			preset = "islands#blueGovernmentCircleIcon"
  			break
  		default:
  			preset = "islands#blueCircleDotIcon"
  		}

  		let newWayPoint = new ymaps.Placemark(this.state.wayPoints[i].coords, {}, { draggable: true, preset, })
  		newWayPoint.properties.set("index", i)
  		newWayPoint.properties.set("name", this.state.wayPoints[i].name)
  		this.state.yaMap.geoObjects.add(newWayPoint)

  		newWayPoint.events.add("dragend", e => {
  			this.onWayPointDrag(e.get("target").properties.get("index"), e.get("target").geometry.getCoordinates())
  		})

  		newWayPoint.events.add("click", e => {
  			this.state.yaMap.balloon.open(e.get("target").geometry.getCoordinates(), {
  				contentHeader: e.get("target").properties.get("name")
            || e.get("target").geometry.getCoordinates().map(it => Math.round(it * 1000000)/1000000).join("; "),
  			})
  		})
  	}

  	if (this.state.wayPoints.length > 1) {
  		let routeLine = new ymaps.Polyline(
      		this.state.wayPoints.map(wayPoint => wayPoint.coords),
  			{},
  			{ strokeWidth: 6, }
  		)
  		this.state.yaMap.geoObjects.add(routeLine)
  	}
  }

  onWayPointAdd = async name => {
  	name = name[0].toUpperCase() + name.slice(1)
  	try {
  		let wayPointToAdd = await ymaps.geocode(name)
  		let coords = wayPointToAdd.geoObjects.get(0).geometry.getCoordinates()
  		this.state.wayPoints.push({ name, coords, })
  		this.state.yaMap.setCenter(coords)
  		this.forceUpdate()
  	} catch (err) {
  		alert("Search error!")
  	}
  }

  onWayPointDrag = async (index, coords) => {
  	this.state.wayPoints[index].coords = coords
  	let newWayPointName
  	try {
  		let reverseGeocoder = await ymaps.geocode(coords, { kind: "house", })
  		newWayPointName = reverseGeocoder.geoObjects.get(0).properties.get("name")
  	} catch (err) {
  		newWayPointName = ""
  	}
  	this.state.wayPoints[index].name = newWayPointName
  	this.forceUpdate()
  }

  onWayPointRemove = index => {
  	this.state.wayPoints.splice(index, 1)
  	this.forceUpdate()
  }

  onWayPointOrderChange = wayPointsUpdated => {
  	this.setState({
  		wayPoints: wayPointsUpdated,
  	})
  }

  renderWayPointListItem = (value, sortIndex) =>
  	<React.Fragment>
    	<div className="waypoint-name">
      	{value.name || value.coords.map(it => Math.round(it * 1000000)/1000000).join("; ")}
    	</div>
    	<button className="btn btn-danger" onClick={() => {this.onWayPointRemove(sortIndex)}}>Remove</button>
  	</React.Fragment>


  componentDidMount() {
  	ymaps.ready(this.init)
  }

  componentDidUpdate() {
  	this.refactorWayPoints()
  }

  render() {
  	return <React.Fragment>
  		<div className="waypoints-list-wrapper">
  			<InputForm onSubmit={this.onWayPointAdd}/>
  			<SortableList
  				items={this.state.wayPoints}
  				onOrderChange={this.onWayPointOrderChange}
  				renderData={this.renderWayPointListItem}
  			/>
  		</div>
  		<div id="map"></div>
  	</React.Fragment>
  }
}
