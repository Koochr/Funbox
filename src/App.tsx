import * as React from 'react';
import SortableList from './components/SortableList';
import InputForm from './components/InputForm';

declare const ymaps: any;

export interface WayPoint {
  name?: string;
  coords: number[];
}

export interface RenderListItem {
  (value: WayPoint, index: number): React.ReactNode;
}

const App: React.FunctionComponent = () => {
  const [wayPoints, setWayPoints] = React.useState<WayPoint[]>([]);
  const [yaMap, setYaMap] = React.useState<any>({});

  const init: () => void = () => {
    setYaMap(
      new ymaps.Map('map', {
        center: [55.76, 37.64],
        zoom: 10,
      }),
    );
  };

  const onWayPointAdd: (name: string) => Promise<void> = async name => {
    name = name[0].toUpperCase() + name.slice(1);
    try {
      const wayPointToAdd = await ymaps.geocode(name);
      const coords = wayPointToAdd.geoObjects.get(0).geometry.getCoordinates();
      setWayPoints([...wayPoints, { name, coords }]);
      yaMap.setCenter(coords);
    } catch (err) {
      alert('Search error!');
    }
  };

  const onWayPointDrag: (index: number, coords: number[]) => Promise<void> = async (index, coords) => {
    let newWayPointName: string;
    try {
      const reverseGeocoder = await ymaps.geocode(coords, { kind: 'house' });
      newWayPointName = reverseGeocoder.geoObjects.get(0).properties.get('name');
    } catch (err) {
      newWayPointName = '';
    }
    setWayPoints(
      wayPoints.map((wayPoint: WayPoint, ind: number) =>
        (ind === index ? { ...wayPoint, name: newWayPointName, coords } : wayPoint)
      )
    );
  };

  const onWayPointRemove: (index: number) => void = index => {
    setWayPoints(wayPoints.filter((wayPoint: WayPoint, ind: number) => ind !== index));
  };

  const refactorWayPoints: () => void = () => {
    if (yaMap.geoObjects) {
      yaMap.geoObjects.removeAll();
      for (let i = 0; i < wayPoints.length; i++) {
        let preset;
        switch (i) {
          case 0:
            preset = 'islands#blueHomeCircleIcon';
            break;
          case wayPoints.length - 1:
            preset = 'islands#blueGovernmentCircleIcon';
            break;
          default:
            preset = 'islands#blueCircleDotIcon';
        }

        const newWayPoint = new ymaps.Placemark(wayPoints[i].coords, {}, { draggable: true, preset });
        newWayPoint.properties.set('index', i);
        newWayPoint.properties.set('name', wayPoints[i].name);
        yaMap.geoObjects.add(newWayPoint);

        newWayPoint.events.add('dragend', (e: any) => {
          onWayPointDrag(e.get('target').properties.get('index'), e.get('target').geometry.getCoordinates());
        });

        newWayPoint.events.add('click', (e: any) => {
          yaMap.balloon.open(e.get('target').geometry.getCoordinates(), {
            contentHeader:
                            e.get('target').properties.get('name')
                            || e
                              .get('target')
                              .geometry.getCoordinates()
                              .map((it: number) => Math.round(it * 1000000) / 1000000)
                              .join('; '),
          });
        });
      }

      if (wayPoints.length > 1) {
        const routeLine = new ymaps.Polyline(
          wayPoints.map((wayPoint: WayPoint) => wayPoint.coords),
          {},
          { strokeWidth: 6 }
        );
        yaMap.geoObjects.add(routeLine);
      }
    }
  };

  const renderWayPointListItem: RenderListItem = (value, sortIndex) => (
    <React.Fragment>
      <div className="waypoint-name">
        {value.name || value.coords.map((it: number) => Math.round(it * 1000000) / 1000000).join('; ')}
      </div>
      <button
        className="btn btn-danger"
        onClick={() => {
          onWayPointRemove(sortIndex);
        }}
      >
          Remove
      </button>
    </React.Fragment>
  );

  React.useEffect(() => {
    ymaps.ready(init);
  }, []);

  React.useEffect(refactorWayPoints);

  return (
    <React.Fragment>
      <div className="waypoints-list-wrapper">
        <InputForm onSubmit={onWayPointAdd} />
        <SortableList items={wayPoints} onOrderChange={setWayPoints} renderData={renderWayPointListItem} />
      </div>
      <div id="map" />
    </React.Fragment>
  );
};

export default App;
