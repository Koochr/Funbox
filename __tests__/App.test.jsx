/* global beforeEach it, jest, expect */
import React from "react"
import { configure, mount, shallow, } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import App from "../react/App.jsx"

configure({ adapter: new Adapter(), })

beforeEach(() => {
	global.ymaps = {
		ready: jest.fn(),
	}
})

it("Should render correctly", () => {
	const component = shallow(
		<App />
	)
	expect(component).toMatchSnapshot()
})

it("Should render needed components", () => {
	const component = mount( <App />)
	component.instance().refactorWayPoints = jest.fn()
	component.setState({
		wayPoints: [
			{
				name: "wp1",
				coords: [5, 10,],
			},
			{
				name: "wp2",
				coords: [10, 20,],
			},
			{
				name: "wp3",
				coords: [50, 100,],
			},
		],
	})
	const wrapper = component.find(".waypoints-list-wrapper")
	expect(wrapper).toHaveLength(1)
	const form = wrapper.find("form")
	expect(form).toHaveLength(1)
	const list = wrapper.find(".list-group")
	expect(list).toHaveLength(1)
	const map = component.find("#map")
	expect(map).toHaveLength(1)
	const names = wrapper.find(".waypoint-name")
	expect(names).toHaveLength(component.state().wayPoints.length)
	const buttons = wrapper.find(".btn-danger")
	expect(buttons).toHaveLength(component.state().wayPoints.length)
	for (let i=0; i<component.state().wayPoints.length; i++) {
		expect (names.at(i).props().children).toStrictEqual(component.state().wayPoints[i].name)
	}
})

it("Should call lifecycle methods", () => {
	const component = shallow( <App />)
	expect(global.ymaps.ready.mock.calls.length).toStrictEqual(1)
	component.instance().refactorWayPoints = jest.fn()
	component.setState({
		wayPoints: ["someNewValue",],
	})
	expect(component.instance().refactorWayPoints.mock.calls.length).toStrictEqual(1)
})

it("Should handle waypoints refactoring on state change", () => {
	const mockPropertiesSet = jest.fn()
	const mockEventsAdd = jest.fn()

	global.ymaps.Placemark = jest.fn().mockImplementation(() => {
		return {
			properties: {
				set: mockPropertiesSet,
			},
			events: {
				add: mockEventsAdd,
			},
		}
	})
	global.ymaps.Polyline = jest.fn().mockImplementation(() => {
		return { mockPolyLine: "mockPolyline", }
	})

	const component = shallow( <App />)
	component.setState({
		wayPoints: [
			{
				name: "wp1",
				coords: [5, 10,],
			},
			{
				name: "",
				coords: [10, 20,],
			},
			{
				name: "wp3",
				coords: [50, 100,],
			},
		],
		yaMap: {
			geoObjects: {
				removeAll: jest.fn(),
				add: jest.fn(),
			},
		},
	})

	expect(component.state().yaMap.geoObjects.removeAll.mock.calls.length).toStrictEqual(1)
	for (let i=0; i<component.state().wayPoints.length; i++) {
		let preset
		switch (i) {
		case 0:
			preset = "islands#blueHomeCircleIcon"
			break
		case component.state().wayPoints.length-1:
			preset = "islands#blueGovernmentCircleIcon"
			break
		default:
			preset = "islands#blueCircleDotIcon"
		}
		expect(global.ymaps.Placemark)
			.toBeCalledWith(component.state().wayPoints[i].coords, {}, { draggable: true, preset, })
		expect(mockPropertiesSet).toBeCalledWith("index", i)
		expect(mockPropertiesSet).toBeCalledWith("name", component.state().wayPoints[i].name)
		expect(component.state().yaMap.geoObjects.add).toBeCalledWith({
			properties: {
				set: mockPropertiesSet,
			},
			events: {
				add: mockEventsAdd,
			},
		})
		expect(mockEventsAdd).toBeCalledWith("dragend", expect.any(Function))
		expect(mockEventsAdd).toBeCalledWith("click", expect.any(Function))
	}

	expect(global.ymaps.Polyline).toBeCalledWith(
		component.state().wayPoints.map(wayPoint => wayPoint.coords),
		{},
		{ strokeWidth: 6, }
	)
	expect(component.state().yaMap.geoObjects.add).toBeCalledWith({ mockPolyLine: "mockPolyline", })
})

it("Should handle waypoint delete button click", () => {
	const component = mount( <App />)
	component.instance().refactorWayPoints = jest.fn()
	const initialWayPoints = [
		{
			name: "wp1",
			coords: [5, 10,],
		},
		{
			name: "wp2",
			coords: [10, 20,],
		},
		{
			name: "wp3",
			coords: [50, 100,],
		},
	]
	const wayPointToRemove = 1
	component.setState({
		wayPoints: initialWayPoints.slice(),
	})
	const button = component.find(".btn-danger").at(wayPointToRemove)
	button.simulate("click")
	expect(component.state().wayPoints)
		.toStrictEqual(initialWayPoints.slice(0, wayPointToRemove).concat(initialWayPoints.slice(wayPointToRemove + 1)))
})

