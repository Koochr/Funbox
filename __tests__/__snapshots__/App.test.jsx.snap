// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="waypoints-list-wrapper"
        >
          <InputForm
            onSubmit={[Function]}
          />
          <WayPointsList
            items={Array []}
            onOrderChange={[Function]}
            renderData={[Function]}
          />
        </div>,
        <div
          id="map"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <InputForm
              onSubmit={[Function]}
            />,
            <WayPointsList
              items={Array []}
              onOrderChange={[Function]}
              renderData={[Function]}
            />,
          ],
          "className": "waypoints-list-wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "items": Array [],
              "onOrderChange": [Function],
              "renderData": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "map",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="waypoints-list-wrapper"
          >
            <InputForm
              onSubmit={[Function]}
            />
            <WayPointsList
              items={Array []}
              onOrderChange={[Function]}
              renderData={[Function]}
            />
          </div>,
          <div
            id="map"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <InputForm
                onSubmit={[Function]}
              />,
              <WayPointsList
                items={Array []}
                onOrderChange={[Function]}
                renderData={[Function]}
              />,
            ],
            "className": "waypoints-list-wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onSubmit": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "items": Array [],
                "onOrderChange": [Function],
                "renderData": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "map",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
